src/lib/io/sol-memmap-storage.c

```diff
@@ -343,19 +361,160 @@ check_map(const struct sol_memmap_map *map)
+
+    memcpy(result, slice->data, slice->len);
+
+    return result;
+}
+
+static int
+resolve_i2c_path(const char *path, struct sol_memmap_map *map)
+{
+    char *rel_path = NULL, *dev_number_s = NULL, *dev_name = NULL, *end_ptr;
+    unsigned int dev_number;
+    struct sol_vector instructions;
+    struct sol_buffer result_path = SOL_BUFFER_INIT_EMPTY;
+    struct sol_str_slice command = sol_str_slice_from_str(path);
+    int ret = -EINVAL;
+
+    instructions = sol_util_str_split(command, ",", 5);
+    if (instructions.len < 5) {
+        SOL_WRN("Invalid create device path. Expected 'create,i2c,<rel_path>,"
+            "<devnumber>,<devname>'");
+        goto end;
+    }
+
+    rel_path = sol_slice_to_str(sol_vector_get(&instructions, REL_PATH_IDX));
+    SOL_NULL_CHECK_GOTO(rel_path, end);
+
+    dev_number_s = sol_slice_to_str(sol_vector_get(&instructions, DEV_NUMBER_IDX));
+    SOL_NULL_CHECK_GOTO(dev_number_s, end);
+
+    errno = 0;
+    dev_number = strtoul(dev_number_s, &end_ptr, 0);
+    if (errno || *end_ptr != '\0')
+        goto end;
+
+    dev_name = sol_slice_to_str(sol_vector_get(&instructions, DEV_NAME_IDX));
+    SOL_NULL_CHECK_GOTO(dev_name, end);
+
+    ret = sol_i2c_create_device(rel_path, dev_name, dev_number,
+        &result_path);
+
+    if (ret >= 0 || ret == -EEXIST) {
+        const struct sol_str_slice ending = SOL_STR_SLICE_LITERAL("/eeprom");
+        struct timespec start;
+        struct stat st;
+
+        ret = sol_buffer_append_slice(&result_path, ending);
+        if (ret < 0)
+            goto end;
+
+        map->path = sol_buffer_steal(&result_path, NULL);
+
+        ret = 0;
+        start = sol_util_timespec_get_current();
+        while (stat(map->path, &st)) {
+            struct timespec elapsed, now = sol_util_timespec_get_current();
+
+            sol_util_timespec_sub(&now, &start, &elapsed);
+            /* Let's wait up to one second */
+            if (elapsed.tv_sec > 0) {
+                ret = -ENODEV;
+                goto end;
+            }
+        }
+    }
+
+end:
+    free(rel_path);
+    free(dev_number_s);
+    free(dev_name);
+    sol_vector_clear(&instructions);
+
+    return ret;
+}
+#endif
+
+static struct sol_memmap_map *
+resolve_map(const struct sol_memmap_map *map)
+{
+#ifdef USE_I2C
+    struct sol_memmap_map *resolved_map = NULL;
+    int r;
+
+    if (strstartswith(map->path, "create,i2c,")) {
+        resolved_map = calloc(1, sizeof(struct sol_memmap_map));
+        SOL_NULL_CHECK(resolved_map, NULL);
+
+        resolved_map->version = map->version;
+        resolved_map->entries = map->entries;
+
+        if (resolve_i2c_path(map->path, resolved_map) < 0) {
+            SOL_WRN("Could not create i2c EEPROM device using command [%s]", map->path);
+            goto error;
+        }
+
+        r = sol_ptr_vector_append(&resolved_maps_to_free, resolved_map);
+        SOL_INT_CHECK_GOTO(r, < 0, error);
+
+        return resolved_map;
+    }
+
+    return (void *)map;
```

Comment:
cast to `(void*)` is smaller to type, but potentially dangerous if we happen to change function signatures.